server/trace/mock_trace_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math/rand"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/empty"
+
+	"google.golang.org/genproto/googleapis/devtools/cloudtrace/v2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+	"google.golang.org/protobuf/proto"
+)
+
+const bufSize = 1024 * 1024
+
+var (
+	client cloudtrace.TraceServiceClient
+	conn   *grpc.ClientConn
+	ctx    context.Context
+	lis    *bufconn.Listener
+)
+
+func setup() {
+	// Setup the in-memory server.
+	lis = bufconn.Listen(bufSize)
+	grpcServer := grpc.NewServer()
+	cloudtrace.RegisterTraceServiceServer(grpcServer, &MockTraceServer{})
+	go func() {
+		if err := grpcServer.Serve(lis); err != nil {
+			log.Fatalf("server exited with error: %v", err)
+		}
+	}()
+
+	// Setup the connection and client.
+	ctx = context.Background()
+	var err error
+	conn, err = grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("failed to dial bufnet: %v", err)
+	}
+	client = cloudtrace.NewTraceServiceClient(conn)
+}
+
+func tearDown() {
+	conn.Close()
```

Comment:
Is this sufficient to close the connection gracefully? Do we not have to call https://godoc.org/google.golang.org/grpc#Server.Stop?

That is, if we do this is it guaranteed that `grpcServer.Serve(lis)` will return nil?

Reply:
Hmm that's true, it says in the docs `Serve will return a non-nil error unless Stop or GracefulStop is called`, so we should definitely call `Stop`, not sure why it wasn't returning an error before. Maybe it calls `Stop` behind the scenes, but I added the call to be safe